---
title: "Network analysis toolkit"
author: "Christopher Conley, Pei Wang, Jie Peng"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Network Analysis Toolkit}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo = FALSE}
library(knitr)
opts_chunk$set(message=F, warning=F)
```

<!--While there exist many general tools for network analysis, it can be difficult to adapt those tools for specific needs of integrative genomic networks. --->
The goal of this vignette is to illustrate the network analysis toolkit in the spaceMap analysis pipeline.  We use the fitted spaceMap networks to the  [bcpls data](https://topherconley.github.io/neta-bcpls/)  as an example. 


<!--
We use the a data set from the Breast Cancer Proteogenomics Landscape Study [BCPLS](http://www.nature.com/nature/journal/v534/n7605/full/nature18003.html). 
Our interest is to use the spaceMap model to find major impacts of DNA copy number alterations (CNA) on protein abundance from 77 breast cancer tumors. The data from the original study has been reduced in dimension and further processed to find the interactions between 1662 CNA (i.e. predictors $\bf x$) and 1595 protein expressions (i.e. responses $\bf y$), as well as the interactions between the protein expressions themselves. The pre-processing and organization of node annotations have been previously completed and is described in more detail on the [neta-bcpls webpage](https://topherconley.github.io/neta-bcpls/) so that the focus of this vignette is on the network analysis. -->

## Input

<!--We take a moment to review  the required (and optional) input data to the network analysis toolkit.  -->
Before conducting the network analysis, you should have completed the model fitting step such that you have a network ensemble or a final network to analyze (e.g. output from [bootVote](https://topherconley.github.io/spacemap/reference/bootVote.html)). 
To fully realize the capability of the toolkit,  you should also have annotations for the nodes. 

Load the `bcpls` list.

```{r}
library(spacemap)
data("bcpls")
#Attach "bcpls" data objects to the R Global Environment
attach(bcpls)
```

Next we list the contents of `bcpls`.

```{r}
names(bcpls)
```


+ The `net` object is a list of two adjacency matrices that is output from [bootVote](https://topherconley.github.io/spacemap/reference/bootVote.html). The protein--protein edges are encoded with 1's in `net$yy`. Likewise, the CNA--protein edges are encoded with 1's in `net$xy`.  

+ The annotation objects `xinfo` and `yinfo` are of type `data.frame`, whose rows should correspond to the rows of `net$xy` and `net$yy`, respectively. <!--If no annotation exists for the nodes $\bf x$ or $\bf y$, then the bare minimum expected is a character vector 'id' for labeling the nodes. -->Let us look at the contents of `yinfo`: 

1. The first column, `id`,  is always required and must be unique for each node; 
2. The second column, `alias`, is optional and may be used for more human-readable labels. For example, two protein isoforms must have different `id` but may have the same gene symbol alias. 
3. If genome coordinates are supplied for cis/trans identification, then we need three additional columns: `chr` a character for the chromosome number, `start` an integer for the starting positiion, and `end` for the ending position.

4. The users can also add other (optional) columns such as `strand` (for DNA strand),  `description` for additional annotations, etc., which will be exported to visulization software such [cytoscrape](http://cytoscape.org/index.html).
5. Finally, missing information should be denoted by 'NA'. 

```{r,eval=FALSE}
head(yinfo,5)
```

```{r, echo=FALSE}
knitr::kable(yinfo[1:5,], row.names = FALSE)
```

Similar rules are applicable for `xinfo`. In this example, `id` contains genomic coordinates, while `alias` contains cytoband information. 


```{r,eval=FALSE}
head(xinfo,5)
```

```{r, echo=FALSE}
knitr::kable(xinfo[1:5,], row.names = FALSE)
```

+ The mapping between the proteins and their  Gene Ontology (GO) terms is  specified in the list `go2eg`: each element is associated with a GO ID and is a character vector specifying all the proteins that map to this GO ID. Proteins can map to multiple GO terms. The mapping `go2eg` will be used for conducting functional enrichment analysis, but is otherwise optional. 

```{r}
head(go2eg,2)
```

<!--The GO ID is useful for an abbreviated mapping, but the actual biological process name is often more informative at the cost of being very verbose.-->
We can also create a biologically informative alias for each GO ID, which will be useful for reporting enriched hubs/modules later. 

```{r}
library(AnnotationDbi)
library(GO.db)
process_alias <- AnnotationDbi::Term(names(go2eg))
head(process_alias,3)
```

+ The list `bdeg` comes from  [bootVote](https://topherconley.github.io/spacemap/reference/bootVote.html). It is optional for network analysis, but would be useful for prioritizing both CNA- and protein- hubs.

1. `bdeg$yy[b,]` is an integer vector representing the degree distribution of the proteins in the network fitted on the $b$th bootstrap replicate.  

2. Similarly, element `bdeg$xy[b,]` is an integer vector representing the degree distribution of the CNAs. 

3. If `bdeg` is provided in the function `rankHub`, then hubs will be prioritized according to their average degree rank, so that highly ranked hubs  would  consistently have a large degree across the network ensemble. 

## Map Annotations

Convert the final network into an *igraph* object and map the annotations onto the network. 

```{r}
library(igraph)
ig <- spacemap::adj2igraph(yy = net$yy, xy = net$xy, yinfo = yinfo, xinfo = xinfo)
```

If we query the attribute names of the nodes in the graph, we notice that the columns of `xinfo` and `yinfo` have been applied. 

```{r}
vertex_attr_names(graph = ig)
```

The igraph package has a number of ways to access the annotation information. For example, if we wish to confirm the chromosome location of ERBB2, we can easily query: 

```{r}
vertex_attr(graph = ig, name = "chr", index = V(ig)[alias %in% "ERBB2"])
```

<!--Our toolkit adds value by quickly summarizing annotations in a format that is publication ready. The next section illustrates how. -->


## Hub Analysis

We first prioritize the CNA- and protein- hubs. If  the `bdeg` argument is specified, then we rank the hubs according to the average degree rank. Accordingly, the most highly ranked hubs will have the most consistently high degree across network ensemble. 

To rank the protein nodes, use the `rankHub` command and simply specify the `level = "y"` argument. 

```{r}
ig <- rankHub(ig = ig, bdeg = bdeg$yy, level = "y")
```

To rank the CNA nodes, specify the `level = "x"` argument. 

```{r}
ig <- rankHub(ig = ig, bdeg = bdeg$xy, level = "x")
```

Alternatively, if the `bdeg` argument is not available, the ranking is according to degree in the final network (coded by "ig"). 

```{r}
tmp <- rankHub(ig = ig,  level = "x")
```

### Identify cis and trans
Next label $x-y$ edges as being regulated in cis or in trans. The *GenomicRanges* package and the genomic coordinates `chr,start,end` columns of `xinfo` and `yinfo` are required for this step. If you don't have the package  [GenomicRanges](http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html) installed yet, try: 

```{r, eval = FALSE}
## try http:// if https:// URLs are not supported
source("https://bioconductor.org/biocLite.R")
biocLite("GenomicRanges")
library(GenomicRanges)
```

Now we can label the $x-y$ edges with either "cis" or "trans" in the `cis_trans` edge attribute of `ig`.

```{r}
ig <- cisTrans(ig = ig, level = "x-y")
```

### Report 
We then report a well-organized table, as seen in Table 3 of the spaceMap publication. The `top` argument allows us to control how many hubs are displayed.

```{r}
xhubs <- reportHubs(ig, top = 6, level = "x")
```

```{r, echo=FALSE}
kable(xhubs, row.names = FALSE)
```

Similarly, we can report the top 10 protein hubs,  their degrees in the  final network, and a description of each hub, if the `description` column was specified in `yinfo`. 

```{r}
yhubs <- reportHubs(ig, top = 10, level = "y")
```

```{r, echo=FALSE}
kable(yhubs, row.names = FALSE)
```

### GO-neighbor percentage
A _CNA neighborhood_ comprises all protein nodes that are directly connected to a CNA hub by an edge. CNA neighborhoods  represent direct perturbations to the proteome by amplifications or deletions in the DNA. To quantify their functional relevance, we compute a score called the _GO-neighbor percentage_. Two protein nodes are called GO-neighbors if they share a common GO term in the same CNA neighborhood. We postulate that a high percentage of GO-neighbors within a CNA neighborhood associates the CNA hub with more functional meaning. These scores, as presented in Figure 5 of the publication, can be generated with a GO mapping to the proteins as follows. 

```{r}
hgp <- xHubEnrich(ig = ig, go2eg = go2eg)
```

```{r, echo=FALSE}
kable(hgp, row.names = FALSE)
```

## Module Analysis

There are many criteria to define modules of a network. This toolkit  allows users to import the module membership information by themselves (see `mods` argument in [modEnrich](https://topherconley.github.io/spacemap/reference/modEnrich.html)).  

In the spaceMap publication, we use the edge-betweenness algorithm in *igraph*.

```{r}
library(igraph)
mods <- cluster_edge_betweenness(ig)
```

The main goal of module analysis is identifying modules that are functionally enriched. 
The `modEnrich` function will test for significantly over-represented GO-terms (or any other valid functional grouping) within a module using hyper-geometric tests.  


In  the current example, only the protein nodes have functional mapping and we specify this through the `levels = "y"` argument. If both predictors and responses have functional mapping in the `go2eg` argument, then we can specify `levels = c("y","x")`.  Other arguments are available to control the enrichment testing; see [modEnrich](https://topherconley.github.io/spacemap/reference/modEnrich.html) for more details. 

```{r}
outmod <- modEnrich(ig = ig, mods = mods, levels = "y", go2eg = go2eg, process_alias = process_alias)
```

The output of the module analysis is a list of 3 objects.

```{r}
names(outmod)
```

+  `ig` is the input igraph network object updated with a "process_id" attribute for nodes affiliated with a significant GO-term. The "process_id" and "module" attributes together are
useful for visualizing which nodes are enriched for a specific biological function. 

+ `etab` is the polished module enrichment table to report significant GO terms, the representation of the GO term in the module relative to the size of the GO term, and which CNA hubs  belong to the module. The top ten hits as in Table S.5 of the spaceMap publication's supplementary materials are as follows: 


```{r, eval = FALSE}
outmod$etab[1:10,]
```


```{r, echo=FALSE}
knitr::kable(outmod$etab[1:10,], row.names = FALSE)
```

+ `eraw` contains details for each (module, GO-term) pair that was subjected to the hyper-geometric test. This output gives the user more control by reporting all tests, the relative over-representation of a GO-term in that module, the raw P-value, and the adjusted P-value. 

```{r, eval = FALSE}
outmod$eraw[1:5,]
```

```{r, echo=FALSE}
knitr::kable(outmod$eraw[1:5,], row.names = FALSE)
```

## Export for Visualization

There are many tools for network visualization. In the publication of spaceMap, we exported the annotated *igraph* network object `ig` to the "graphml" format, which maintained all the attributes associated with nodes when read into *Cytoscape* for visualization. 

```{r}
igraph::write_graph(graph = outmod$ig, file = tempfile(), format = "graphml")
```

Here we list all the attributes associated with the nodes that can be used in tandem with Cytoscape's filtering functions to identify specific nodes of interest. 

```{r}
vertex_attr_names(outmod$ig)
```

We describe some of the most useful attributes for visualization:

+ 'name': the unique node ID 
+ 'alias': the node alias (e.g. gene symbol ERBB2)
+ 'chr,start,end,strand': the gene coordinates of nodes
+ 'description': any note (e.g. breast cancer oncogene)
+ 'levels': indicates whether the node belongs to predictors "x" or responses "y"
+ 'rank_hub': the rank of the hub within its level (e.g. a value of "1" for a node of level "x" means that it is the most consistently high degree $x$ node in the network. ) 
+ 'regulates_in_cis': list of genes regulated in cis 
+ 'module': the module ID that the node belongs to. 
+ 'process_id': the significant GO-term(s) associated with the node. 


Also the edge attributes are exported to 'graphml' format.

```{r}
edge_attr_names(outmod$ig)
```

+ 'levels' indicates whether an edge is $x-y$ or $y-y$.
+ 'cis_trans' indicates whether an edge is regulated in cis or in trans. 

## Summary

This toolkit is useful for analyzing and summarizing the output of  the spaceMap model fit in the context of integrative genomics. The biological context mapped onto the network object can easily be exported to  existing network visualization tools like *Cytoscape*. 


