<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Network analysis toolkit • spacemap</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">spacemap</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://github.com/topherconley/spacemap/releases/tag/v0.45.0-beta">Installation</a>
</li>
<li>
  <a href="../reference/index.html">Docs</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basics.html">Model Fitting Basics</a>
    </li>
    <li>
      <a href="../articles/tuning.html">Model Tuning</a>
    </li>
    <li>
      <a href="../articles/ensemble.html">Ensemble Network</a>
    </li>
    <li>
      <a href="../articles/neta.html">Network Analysis Toolkit</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://topherconley.github.io/neta-bcpls/">Application</a>
</li>
<li>
  <a href="https://github.com/topherconley/spacemap/issues">Contact</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/topherconley/spacemap">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Network analysis toolkit</h1>
                        <h4 class="author">Christopher Conley, Pei Wang, Jie Peng</h4>
            
            <h4 class="date">2017-05-08</h4>
          </div>

    
    
<div class="contents">
<!--While there exist many general tools for network analysis, it can be difficult to adapt those tools for specific needs of integrative genomic networks. --->
<p>The goal of this vignette is to illustrate the network analysis toolkit in the spaceMap analysis pipeline. We use the fitted spaceMap networks to the <code>bcpls</code> data <a href="https://topherconley.github.io/neta-bcpls/">neta-bcpls webpage</a> as an example.</p>
<!--
We use the a data set from the Breast Cancer Proteogenomics Landscape Study [BCPLS](http://www.nature.com/nature/journal/v534/n7605/full/nature18003.html). 
Our interest is to use the spaceMap model to find major impacts of DNA copy number alterations (CNA) on protein abundance from 77 breast cancer tumors. The data from the original study has been reduced in dimension and further processed to find the interactions between 1662 CNA (i.e. predictors $\bf x$) and 1595 protein expressions (i.e. responses $\bf y$), as well as the interactions between the protein expressions themselves. The pre-processing and organization of node annotations have been previously completed and is described in more detail on the [neta-bcpls webpage](https://topherconley.github.io/neta-bcpls/) so that the focus of this vignette is on the network analysis. -->
<div id="input" class="section level1">
<h1 class="hasAnchor">
<a href="#input" class="anchor"></a>Input</h1>
<!--We take a moment to review  the required (and optional) input data to the network analysis toolkit.  -->
<p>Before conducting the network analysis, you should have completed the model fitting step such that you have a network ensemble or one final network to analyze (e.g. output from <a href="https://topherconley.github.io/spacemap/reference/bootVote.html">bootVote</a>). To fully realize the capability of the toolkit, you also should have node annotations mapped onto the network.</p>
<p>Now we load the <code>bcpls</code> data object and outline the required input for the toolkit.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(spacemap)
<span class="kw">data</span>(<span class="st">"bcpls"</span>)
<span class="co">#convert list to global environment</span>
.GlobalEnv &lt;-<span class="st"> </span><span class="kw">list2env</span>(<span class="dt">x=</span> bcpls, <span class="dt">envir =</span> <span class="kw">globalenv</span>())
<span class="kw">ls</span>()</code></pre></div>
<pre><code>## character(0)</code></pre>
<p>The <code>net</code> object is a list of two adjacency matrices that is output from <a href="https://topherconley.github.io/spacemap/reference/bootVote.html">bootVote</a>. The protein–protein edges are encoded with 1’s in <code>net$yy</code>. Likewise, the CNA–protein edges are encoded with 1’s in <code>net$xy</code>.</p>
<p>The annotation objects <code>xinfo</code> and <code>yinfo</code> are of type <code>data.frame</code>, whose rows should correspond to the rows of <code>net$xy</code> and <code>net$yy</code>, respectively. <!--If no annotation exists for the nodes $\bf x$ or $\bf y$, then the bare minimum expected is a character vector 'id' for labeling the nodes. --> Let us look at the contents of <code>yinfo</code>: The first column, <code>id</code>, is always required and must be unique for each node; The second column, <code>alias</code>, is optional and may be used for more human-readable labels. For example, two protein isoforms must have different ID’s but may have the same gene symbol alias. If genome coordinates are supplied for cis/trans identification, then we need three additional columns: <code>chr</code> a character for the chromosome number, <code>start</code> an integer for the beginning of the chromsome location, and <code>end</code> for the end of the chromsome location. The users can also add other (optional) columns such as <code>strand</code> (for DNA strand), <code>description</code> for additional annotations, etc., which will be exported to visulization software such cytoscrape. Finally, missing information should be denoted by ‘NA’.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(yinfo[<span class="dv">1</span>:<span class="dv">5</span>,], <span class="dt">row.names =</span> <span class="ot">FALSE</span>)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">id</th>
<th align="left">alias</th>
<th align="left">chr</th>
<th align="right">start</th>
<th align="right">end</th>
<th align="left">strand</th>
<th align="left">description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">NP_112598</td>
<td align="left">EPPK1</td>
<td align="left">chr8</td>
<td align="right">144939491</td>
<td align="right">144952632</td>
<td align="left">-</td>
<td align="left">epiplakin</td>
</tr>
<tr class="even">
<td align="left">NP_001714</td>
<td align="left">DST</td>
<td align="left">chr6</td>
<td align="right">56322784</td>
<td align="right">56819426</td>
<td align="left">-</td>
<td align="left">bullous pemphigoid antigen 1 isoform 1e precursor</td>
</tr>
<tr class="odd">
<td align="left">NP_002465</td>
<td align="left">MYH11</td>
<td align="left">chr16</td>
<td align="right">15796991</td>
<td align="right">15950887</td>
<td align="left">-</td>
<td align="left">myosin-11 isoform SM1A</td>
</tr>
<tr class="even">
<td align="left">NP_001035202</td>
<td align="left">MYH11</td>
<td align="left">chr16</td>
<td align="right">15796991</td>
<td align="right">15950887</td>
<td align="left">-</td>
<td align="left">myosin-11 isoform SM2B</td>
</tr>
<tr class="odd">
<td align="left">NP_001070654</td>
<td align="left">MYH14</td>
<td align="left">chr19</td>
<td align="right">50706884</td>
<td align="right">50813801</td>
<td align="left">+</td>
<td align="left">myosin-14 isoform 1</td>
</tr>
</tbody>
</table>
<p>The same column and row rules are applicable for <code>xinfo</code>. Here, <code>id</code> contains precise genomic coordinates, while the <code>alias</code> contains cytoband information.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(xinfo[<span class="dv">1</span>:<span class="dv">5</span>,], <span class="dt">row.names =</span> <span class="ot">FALSE</span>)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">id</th>
<th align="left">alias</th>
<th align="left">chr</th>
<th align="right">start</th>
<th align="right">end</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1_3218610_6972307_0</td>
<td align="left">1p36.31-36.32</td>
<td align="left">chr1</td>
<td align="right">3218610</td>
<td align="right">6972307</td>
</tr>
<tr class="even">
<td align="left">1_6990271_14007849_0</td>
<td align="left">1p36.21-36.31</td>
<td align="left">chr1</td>
<td align="right">6990271</td>
<td align="right">14007849</td>
</tr>
<tr class="odd">
<td align="left">1_14028633_28692959_0</td>
<td align="left">1p35.3-36.21</td>
<td align="left">chr1</td>
<td align="right">14028633</td>
<td align="right">28692959</td>
</tr>
<tr class="even">
<td align="left">1_28695597_30387702_0</td>
<td align="left">1p35.2-35.3</td>
<td align="left">chr1</td>
<td align="right">28695597</td>
<td align="right">30387702</td>
</tr>
<tr class="odd">
<td align="left">1_30392887_31908168_0</td>
<td align="left">1p35.2</td>
<td align="left">chr1</td>
<td align="right">30392887</td>
<td align="right">31908168</td>
</tr>
</tbody>
</table>
<p>The mapping between the proteins and their biological process term in the Gene Ontology (GO) is specified in the list <code>go2eg</code>, which has named each element with a GO ID; and each element is a character vector specifying all the proteins that map to this GO ID. Proteins can map to multiple GO IDs. The mapping <code>go2eg</code> will be used for conducting functional enrichment analysis of CNA-hubs and network modules, but is otherwise optional.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(go2eg,<span class="dv">2</span>)</code></pre></div>
<pre><code>## $`GO:0000122`
##  [1] "NP_009330"    "NP_644671"    "NP_001120"    "NP_056082"   
##  [5] "NP_001041666" "NP_005321"    "NP_055398"    "NP_036233"   
##  [9] "NP_001441"    "NP_001257972" "NP_001177666" "NP_002492"   
## [13] "NP_005587"    "NP_061119"    "NP_002022"    "NP_060733"   
## [17] "NP_005310"    "NP_005312"    "NP_005311"    "NP_005313"   
## [21] "NP_006051"    "NP_001002295" "NP_116027"    "NP_005987"   
## [25] "NP_005985"    "NP_066285"    "NP_001138627" "NP_073576"   
## [29] "NP_001001548" "NP_004022"    "NP_000336"    "NP_001963"   
## [33] "NP_005323"    "NP_004771"    "NP_001744"    "NP_006262"   
## [37] "NP_620590"   
## 
## $`GO:0000165`
##  [1] "NP_001020029" "NP_000338"    "NP_003117"    "NP_006149"   
##  [5] "NP_004439"    "NP_005219"    "NP_005226"    "NP_003740"   
##  [9] "NP_001356"    "NP_005535"    "NP_057726"    "NP_001553"   
## [13] "NP_000213"    "NP_001744"    "NP_000791"    "NP_620590"</code></pre>
<!--The GO ID is useful for an abbreviated mapping, but the actual biological process name is often more informative at the cost of being very verbose.-->
<p>We can also create a biologically informative alias for each GO ID that will be useful for reporting enriched hubs/modules later.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(AnnotationDbi)
<span class="kw">library</span>(GO.db)
process_alias &lt;-<span class="st"> </span>AnnotationDbi::<span class="kw">Term</span>(<span class="kw">names</span>(go2eg))
<span class="kw">head</span>(process_alias,<span class="dv">3</span>)</code></pre></div>
<pre><code>##                                                             GO:0000122 
## "negative regulation of transcription from RNA polymerase II promoter" 
##                                                             GO:0000165 
##                                                         "MAPK cascade" 
##                                                             GO:0000278 
##                                                   "mitotic cell cycle"</code></pre>
<p>The list <code>bdeg</code> comes from <a href="https://topherconley.github.io/spacemap/reference/bootVote.html">bootVote</a>, which is optional for network analysis, but would be useful for prioritizing both CNA- and protein- hubs. <code>bdeg$yy[b,]</code> is an integer vector representing the degree distribution of the proteins in the network fitted on the <span class="math inline">\(b\)</span>th bootstrap replicate. Similarly, element <code>bdeg$xy[b,]</code> is an integer vector representing the degree distribution of the CNAs. If <code>bdeg</code> is provided in the function “rankHub”, the CNA-hubs will be prioritized according to their average degree rank across the <span class="math inline">\(B\)</span> bootstrap replicates, so that highly ranked hubs would consistently have a large degree across the network ensemble.</p>
</div>
<div id="map-annotations" class="section level1">
<h1 class="hasAnchor">
<a href="#map-annotations" class="anchor"></a>Map Annotations</h1>
<p>Convert the final network into an <em>igraph</em> object and map the annotations onto the network.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(igraph)
ig &lt;-<span class="st"> </span>spacemap::<span class="kw"><a href="../reference/adj2igraph.html">adj2igraph</a></span>(<span class="dt">yy =</span> net$yy, <span class="dt">xy =</span> net$xy, <span class="dt">yinfo =</span> yinfo, <span class="dt">xinfo =</span> xinfo)</code></pre></div>
<p>If we query the attribute names of the nodes in the graph, we notice that the columns of <code>xinfo</code> and <code>yinfo</code> have been applied.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">vertex_attr_names</span>(<span class="dt">graph =</span> ig)</code></pre></div>
<pre><code>## [1] "name"        "alias"       "chr"         "start"       "end"        
## [6] "strand"      "description" "levels"</code></pre>
<p>The igraph package has a number of ways to access the annotation information. For example, if we wish to confirm the chromosome location of ERBB2, we can easily query:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">vertex_attr</span>(<span class="dt">graph =</span> ig, <span class="dt">name =</span> <span class="st">"chr"</span>, <span class="dt">index =</span> <span class="kw">V</span>(ig)[alias %in%<span class="st"> "ERBB2"</span>])</code></pre></div>
<pre><code>## [1] "chr17"</code></pre>
<p>Our toolkit adds value by quickly summarizing annotations in a format that is publication ready. The next section illustrates how.</p>
</div>
<div id="hub-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#hub-analysis" class="anchor"></a>Hub Analysis</h1>
<p>We first prioritize the CNA- and protein- hubs. If the degree distributions from the network ensemble are available in the <code>bdeg</code> argument, then we rank the hubs according to the average degree rank. Accordingly, the most highly ranked hubs will have the most consistently high degree across network ensemble.</p>
<p>To rank the protein nodes, use the <code>rankHub</code> command and simply specify the <code>level = "y"</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ig &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rankHub.html">rankHub</a></span>(<span class="dt">ig =</span> ig, <span class="dt">bdeg =</span> bdeg$yy, <span class="dt">level =</span> <span class="st">"y"</span>)</code></pre></div>
<p>To rank the CNA nodes, specify the <code>level = "x"</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ig &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rankHub.html">rankHub</a></span>(<span class="dt">ig =</span> ig, <span class="dt">bdeg =</span> bdeg$xy, <span class="dt">level =</span> <span class="st">"x"</span>)</code></pre></div>
<p>Alternatively, if the <code>bdeg</code> argument is not available, the ranking is according to degree in the final network (coded by “ig”).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tmp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rankHub.html">rankHub</a></span>(<span class="dt">ig =</span> ig,  <span class="dt">level =</span> <span class="st">"x"</span>)</code></pre></div>
<div id="identify-cis-and-trans" class="section level2">
<h2 class="hasAnchor">
<a href="#identify-cis-and-trans" class="anchor"></a>Identify cis and trans</h2>
<p>Next label <span class="math inline">\(x-y\)</span> edges as being regulated in cis or in trans. The <em>GenomicRanges</em> package and the genomic coordinates <code>chr,start,end</code> columns of <code>xinfo</code> and <code>yinfo</code> are required for this step.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GenomicRanges)
ig &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cisTrans.html">cisTrans</a></span>(<span class="dt">ig =</span> ig, <span class="dt">level =</span> <span class="st">"x-y"</span>)</code></pre></div>
</div>
<div id="report" class="section level2">
<h2 class="hasAnchor">
<a href="#report" class="anchor"></a>Report</h2>
<p>We then report a well-organized table, as seen in Table 3 of the spaceMap publication. The <code>top</code> argument allows us to control how many hubs are displayed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">xhubs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/reportHubs.html">reportHubs</a></span>(ig, <span class="dt">top =</span> <span class="dv">6</span>, <span class="dt">level =</span> <span class="st">"x"</span>)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">hub</th>
<th align="left"># cis/ # trans</th>
<th align="right">Potential # cis</th>
<th align="left">cis genes</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">17q21.32 (46-46 Mb)</td>
<td align="left">1 / 98</td>
<td align="right">14</td>
<td align="left">LRRC59</td>
</tr>
<tr class="even">
<td align="left">5q34 (160-170 Mb)</td>
<td align="left">0 / 129</td>
<td align="right">0</td>
<td align="left">–</td>
</tr>
<tr class="odd">
<td align="left">17q12 (38-38 Mb)</td>
<td align="left">5 / 47</td>
<td align="right">34</td>
<td align="left">ERBB2, GRB7, KAT2A, PNMT, MIEN1</td>
</tr>
<tr class="even">
<td align="left">10p15.1-15.3 (0.42-4 Mb)</td>
<td align="left">0 / 86</td>
<td align="right">10</td>
<td align="left">–</td>
</tr>
<tr class="odd">
<td align="left">15q13.1-15.1 (29-43 Mb)</td>
<td align="left">2 / 34</td>
<td align="right">20</td>
<td align="left">IVD, JMJD7</td>
</tr>
<tr class="even">
<td align="left">16q22.1-22.2 (68-71 Mb)</td>
<td align="left">2 / 70</td>
<td align="right">13</td>
<td align="left">CDH1, CYB5B</td>
</tr>
</tbody>
</table>
<p>Similarly, we can report the top 10 protein hubs, as well as the final network degree, and a description of each hub, if the <code>description</code> column was specified in <code>yinfo</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">yhubs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/reportHubs.html">reportHubs</a></span>(ig, <span class="dt">top =</span> <span class="dv">10</span>, <span class="dt">level =</span> <span class="st">"y"</span>)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">hub</th>
<th align="right">degree</th>
<th align="left">description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ARGLU1</td>
<td align="right">5</td>
<td align="left">arginine and glutamate-rich protein 1</td>
</tr>
<tr class="even">
<td align="left">SERPINA1</td>
<td align="right">8</td>
<td align="left">alpha-1-antitrypsin precursor</td>
</tr>
<tr class="odd">
<td align="left">KLKB1</td>
<td align="right">5</td>
<td align="left">plasma kallikrein preproprotein</td>
</tr>
<tr class="even">
<td align="left">MMP2</td>
<td align="right">5</td>
<td align="left">72 kDa type IV collagenase isoform a preproprotein</td>
</tr>
<tr class="odd">
<td align="left">C6</td>
<td align="right">4</td>
<td align="left">complement component C6 precursor</td>
</tr>
<tr class="even">
<td align="left">APOA1</td>
<td align="right">6</td>
<td align="left">apolipoprotein A-I preproprotein</td>
</tr>
<tr class="odd">
<td align="left">HPX</td>
<td align="right">4</td>
<td align="left">hemopexin precursor</td>
</tr>
<tr class="even">
<td align="left">ALB</td>
<td align="right">5</td>
<td align="left">serum albumin preproprotein</td>
</tr>
<tr class="odd">
<td align="left">C5</td>
<td align="right">3</td>
<td align="left">complement C5 preproprotein</td>
</tr>
<tr class="even">
<td align="left">CFI</td>
<td align="right">3</td>
<td align="left">complement factor I preproprotein</td>
</tr>
</tbody>
</table>
</div>
<div id="go-neighbor-percentage" class="section level2">
<h2 class="hasAnchor">
<a href="#go-neighbor-percentage" class="anchor"></a>GO-neighbor Percentage</h2>
<p>A CNA neighborhood comprises all protein nodes that are directly connected to a CNA hub by an edge. CNA neighborhoods represent direct perturbations to the proteome by amplifications or deletions in the DNA. To quantify their functional relevance, we compute a score called the <em>GO-neighbor</em> percentage. Two protein nodes are called GO-neighbors if they share a common GO term in the same CNA neighborhood. We postulate that a high percentage of GO-neighbors within a CNA neighborhood associates the CNA hub with more functional meaning. These scores, as presented in figure 5 of the publication, can be generated with a GO mapping to the proteins as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hgp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/xHubEnrich.html">xHubEnrich</a></span>(<span class="dt">ig =</span> ig, <span class="dt">go2eg =</span> go2eg)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">hub</th>
<th align="right">degree</th>
<th align="right">neighbor_percentage</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">10p15.1-15.3</td>
<td align="right">86</td>
<td align="right">77.90698</td>
</tr>
<tr class="even">
<td align="left">11q13.5-14.1</td>
<td align="right">11</td>
<td align="right">45.45455</td>
</tr>
<tr class="odd">
<td align="left">12q21.1</td>
<td align="right">4</td>
<td align="right">50.00000</td>
</tr>
<tr class="even">
<td align="left">15q13.1-15.1</td>
<td align="right">36</td>
<td align="right">63.88889</td>
</tr>
<tr class="odd">
<td align="left">16p12.1-12.3</td>
<td align="right">21</td>
<td align="right">71.42857</td>
</tr>
<tr class="even">
<td align="left">16q22.1</td>
<td align="right">16</td>
<td align="right">43.75000</td>
</tr>
<tr class="odd">
<td align="left">16q22.1-22.2</td>
<td align="right">72</td>
<td align="right">79.16667</td>
</tr>
<tr class="even">
<td align="left">17p11.2</td>
<td align="right">7</td>
<td align="right">71.42857</td>
</tr>
<tr class="odd">
<td align="left">17p11.2</td>
<td align="right">5</td>
<td align="right">60.00000</td>
</tr>
<tr class="even">
<td align="left">17q12</td>
<td align="right">52</td>
<td align="right">78.84615</td>
</tr>
<tr class="odd">
<td align="left">17q21.32</td>
<td align="right">99</td>
<td align="right">72.72727</td>
</tr>
<tr class="even">
<td align="left">17q23.1-23.2</td>
<td align="right">21</td>
<td align="right">33.33333</td>
</tr>
<tr class="odd">
<td align="left">5q34</td>
<td align="right">129</td>
<td align="right">77.51938</td>
</tr>
<tr class="even">
<td align="left">5p12-5q11.1</td>
<td align="right">25</td>
<td align="right">60.00000</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="module-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#module-analysis" class="anchor"></a>Module Analysis</h1>
<p>There are many criteria to define modules of a network. This toolkit does not require a specific algorithm for finding modules and allows users to import the module membership information (see <code>mods</code> argument in <a href="https://topherconley.github.io/spacemap/reference/modAnalysis.html">modAnalysis</a>). In the spaceMap publication, we use the edge-betweenness algorithm in <em>igraph</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mods &lt;-<span class="st"> </span><span class="kw">cluster_edge_betweenness</span>(ig)</code></pre></div>
<p>The main goal of module analysis is identifying modules that are functionally enriched. The <code>modAnalysis</code> function will test for significantly over-represented GO-terms (or any other valid functional grouping) in a module using hyper-geometric tests. In this application, only the protein nodes have a functional mapping and we specify this through the <code>levels = "y"</code> argument. If both predictors and responses have a functional mapping in the <code>go2eg</code> argument, then we can specify <code>levels = c("y","x")</code>. Other arguments are available to control the enrichment testing; see the docs of <a href="https://topherconley.github.io/spacemap/reference/modAnalysis.html">modAnalysis</a> for more details.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">outmod &lt;-<span class="st"> </span><span class="kw"><a href="../reference/modAnalysis.html">modAnalysis</a></span>(<span class="dt">ig =</span> ig, <span class="dt">mods =</span> mods, <span class="dt">levels =</span> <span class="st">"y"</span>, <span class="dt">go2eg =</span> go2eg, <span class="dt">process_alias =</span> process_alias)</code></pre></div>
<p>The output of the module analysis is a list of 3 objects.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(outmod)</code></pre></div>
<pre><code>## [1] "ig"   "etab" "eraw"</code></pre>
<ul>
<li><p><code>ig</code> is the input igraph network object updated with a “process_id” attribute for nodes affiliated with a significant GO-term. The “process_id” and “module” attributes together are useful for visualizing which nodes are enriched for a specific biological function.</p></li>
<li><p><code>etab</code> is the polished module enrichment table to report significant GO terms, the representation of the GO term in the module relative to the size of the GO term, and which CNA hubs belong to the module. The top ten hits appear as follows as in Table S.5 of the spaceMap publication’s supplementary materials.</p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(outmod$etab[<span class="dv">1</span>:<span class="dv">10</span>,], <span class="dt">row.names =</span> <span class="ot">FALSE</span>)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">Module (size)</th>
<th align="left">GO Category</th>
<th align="left">GO Obs./ Total</th>
<th align="left">X hubs (hits)</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">M1 (244)</td>
<td align="left">small molecule metabolic process</td>
<td align="left">61/280</td>
<td align="left">5q34 (111); 10p15.1-15.3 (79)</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">oxidation-reduction process</td>
<td align="left">28/106</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">transcription from RNA polymerase II promoter</td>
<td align="left">9/20</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">protein homotetramerization</td>
<td align="left">9/20</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">M10 (28)</td>
<td align="left">transcription, DNA-templated</td>
<td align="left">11/77</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">regulation of transcription, DNA-templated</td>
<td align="left">7/41</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">M2 (25)</td>
<td align="left">negative regulation of cell growth</td>
<td align="left">6/22</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">muscle organ development</td>
<td align="left">5/17</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">M3 (57)</td>
<td align="left">cell proliferation</td>
<td align="left">6/37</td>
<td align="left">17q12 (50)</td>
</tr>
<tr class="even">
<td align="left">M4 (97)</td>
<td align="left">ion transmembrane transport</td>
<td align="left">10/21</td>
<td align="left">17q21.32 (79)</td>
</tr>
</tbody>
</table>
<ul>
<li>
<code>eraw</code> contains details for each (module, GO-term) pair that was subjected to the hyper-geometric test. This output gives the user more control by reporting all tests, the relative over-representation of a GO-term in that module, the raw P-value, and the adjusted P-value.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(outmod$eraw[<span class="dv">1</span>:<span class="dv">5</span>,], <span class="dt">row.names =</span> <span class="ot">FALSE</span>)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">Module</th>
<th align="left">process_id</th>
<th align="left">process_alias</th>
<th align="left">GO Obs./Total</th>
<th align="right">P_value</th>
<th align="right">Adjusted_P_value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">M1</td>
<td align="left"><a href="GO:0044281" class="uri">GO:0044281</a></td>
<td align="left">small molecule metabolic process</td>
<td align="left">61/280</td>
<td align="right">0.0002922</td>
<td align="right">0.0139429</td>
</tr>
<tr class="even">
<td align="left">M1</td>
<td align="left"><a href="GO:0055114" class="uri">GO:0055114</a></td>
<td align="left">oxidation-reduction process</td>
<td align="left">28/106</td>
<td align="right">0.0008225</td>
<td align="right">0.0327023</td>
</tr>
<tr class="odd">
<td align="left">M1</td>
<td align="left"><a href="GO:0006366" class="uri">GO:0006366</a></td>
<td align="left">transcription from RNA polymerase II promoter</td>
<td align="left">9/20</td>
<td align="right">0.0010725</td>
<td align="right">0.0398029</td>
</tr>
<tr class="even">
<td align="left">M1</td>
<td align="left"><a href="GO:0051289" class="uri">GO:0051289</a></td>
<td align="left">protein homotetramerization</td>
<td align="left">9/20</td>
<td align="right">0.0010725</td>
<td align="right">0.0398029</td>
</tr>
<tr class="odd">
<td align="left">M1</td>
<td align="left"><a href="GO:0034641" class="uri">GO:0034641</a></td>
<td align="left">cellular nitrogen compound metabolic process</td>
<td align="left">17/61</td>
<td align="right">0.0050060</td>
<td align="right">0.1248049</td>
</tr>
</tbody>
</table>
</div>
<div id="export-for-network-visualization" class="section level1">
<h1 class="hasAnchor">
<a href="#export-for-network-visualization" class="anchor"></a>Export for Network Visualization</h1>
<p>There exist many tools for network visualization. In the publication of spaceMap, we exported the annotated <em>igraph</em> network object <code>ig</code> to the “graphml” format, which maintained all the attributes associated with nodes when read into <em>Cytoscape</em> for visualization.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">igraph::<span class="kw">write_graph</span>(<span class="dt">graph =</span> outmod$ig, <span class="dt">file =</span> <span class="kw">tempfile</span>(), <span class="dt">format =</span> <span class="st">"graphml"</span>)</code></pre></div>
<p>Here we list all the attributes associated with the nodes (i.e. vertices) that can be used in tandem with Cytoscape’s filtering functions to identify specific nodes of interest.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">vertex_attr_names</span>(outmod$ig)</code></pre></div>
<pre><code>##  [1] "name"             "alias"            "chr"             
##  [4] "start"            "end"              "strand"          
##  [7] "description"      "levels"           "rank_hub"        
## [10] "mean_rank_hub"    "sd_rank_hub"      "ncis"            
## [13] "ntrans"           "regulates_in_cis" "potential_cis"   
## [16] "module"           "process_id"</code></pre>
<p>We describe some of the most useful attributes for visualization.</p>
<ul>
<li>‘name’: the unique node ID</li>
<li>‘alias’: the node alias (e.g. gene symbol ERBB2)</li>
<li>‘chr,start,end,strand’: the gene coordinates of nodes</li>
<li>‘description’: any note (e.g. breast cancer oncogene)</li>
<li>‘levels’: indicates the node belongs to predictors “x” or response level “y”</li>
<li>‘rank_hub’: the rank of the hub within level (e.g. a value of “1” for a node of level x implies the most consistently high degree <span class="math inline">\(x\)</span> node in the network. )</li>
<li>‘regulates_in_cis’: list of genes regulated in cis</li>
<li>‘module’: the module ID the node belongs to.</li>
<li>‘process_id’: the significant GO-term(s) associated with the node.</li>
</ul>
<p>Also the edge attributes are exported to ‘graphml’ format.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">edge_attr_names</span>(outmod$ig)</code></pre></div>
<pre><code>## [1] "levels"    "cis_trans"</code></pre>
<ul>
<li>‘levels’ indicates an edge is <span class="math inline">\(x-y\)</span> or <span class="math inline">\(y-y\)</span>
</li>
<li>‘cis_trans’ indicates an edge is either regulated cis or trans.</li>
</ul>
</div>
<div id="summary" class="section level1">
<h1 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h1>
<p>This toolkit is useful and convenient for analyzing and summarizing the output of spaceMap in the context of integrative genomics. The biological context mapped onto the network object can easily be exported to existing network visualization tools like <em>Cytoscape</em>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#input">Input</a></li>
      <li><a href="#map-annotations">Map Annotations</a></li>
      <li>
<a href="#hub-analysis">Hub Analysis</a><ul class="nav nav-pills nav-stacked">
<li><a href="#identify-cis-and-trans">Identify cis and trans</a></li>
      <li><a href="#report">Report</a></li>
      <li><a href="#go-neighbor-percentage">GO-neighbor Percentage</a></li>
      </ul>
</li>
      <li><a href="#module-analysis">Module Analysis</a></li>
      <li><a href="#export-for-network-visualization">Export for Network Visualization</a></li>
      <li><a href="#summary">Summary</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/topherconley/">Christopher Conley</a>, <a href="http://research.mssm.edu/wanglab/index.htm">Pei Wang</a>, <a href="http://www.stat.ucdavis.edu/~jie/">Jie Peng</a>, UC Davis.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
